src/Hazelcast.Net/Core/SocketExtensions.cs

```diff
@@ -49,50 +49,59 @@ public static async Task ConnectAsync(this Socket socket, EndPoint endPoint, int
-            e.RemoteEndPoint = endPoint;
-            e.Completed += ConnectCallback;
-            var pending = socket.ConnectAsync(e);
+            static void CompleteCallback(object sender, SocketAsyncEventArgs sea)
+                => Complete(sea, (TaskCompletionSource<Socket>) sea.UserToken);
 
-            Task task;
-            if (pending)
+            var connected = new TaskCompletionSource<Socket>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            using var sea = new SocketAsyncEventArgs
             {
-                var timeoutCancel = new CancellationTokenSource();
-                var timeoutTask = Task.Delay(timeoutMilliseconds, timeoutCancel.Token);
-                var reg = cancellationToken.Register(() => timeoutCancel.Cancel());
-
-                task = await Task.WhenAny(connected.Task, timeoutTask).CfAwait();
-                await reg.DisposeAsync().CfAwait();
-                timeoutCancel.Cancel();
-                timeoutCancel.Dispose();
-                await timeoutTask.CfAwaitNoThrow();
-            }
-            else
+                UserToken = connected,
+                RemoteEndPoint = endPoint
+            };
+
+            sea.Completed += CompleteCallback;
+            var pending = socket.ConnectAsync(sea);
+
+            if (!pending)
             {
-                task = connected.Task;
-                ConnectCallback(null, e);
+                // "The I/O operation completed synchronously. In this case, The Completed event on the
+                // sea parameter will not be raised and the sea object passed as a parameter may be
+                // examined immediately"
+                if (sea.SocketError != SocketError.Success)
+                    throw new SocketException((int) sea.SocketError);
+                return;
             }
 
-            e.Dispose();
+            // "the I/O operation is pending. The Completed event on the sea parameter will be raised
+            // upon completion of the operation."
+            // but, we don't want to wait forever, and have to deal with timeout + cancellation.
+
+            using var timeoutCancel = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            var timeoutTask = Task.Delay(timeoutMilliseconds, timeoutCancel.Token);
+            var task = await Task.WhenAny(connected.Task, timeoutTask).CfAwait();
+
+            if (!timeoutTask.IsCompletedSuccessfully())
+            {
+                timeoutCancel.Cancel(); // don't leave the timeoutTask running
+                await timeoutTask.CfAwaitNoThrow(); // and make sure it does not leak unobserved exception
+            }
 
-            if (task == connected.Task)
+            if (task != timeoutTask)
             {
-                await task.CfAwait(); // throw if there is anything to throw (failed to connect...)
+                connected.Task.GetAwaiter().GetResult(); // we know the task has completed - throw if needed
```

Comment:
Here is `GetAwaiter()` called because `await` can introduce unnecessary state machine on a completed task, right?

Reply:
correct